ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1948,6 +1952,49 @@ void shutdown() throws InterruptedException {
     shutdownLatch.await();
   }
 
+  /**
+   * This class is used to generate remote replica groups continuously for replication
+   * TODO -fill more details as we keep writing code
+   */
+  class RemoteReplicaGroupPoller {
+    private final List<DataNodeTracker> dataNodeTrackers;
+
+    RemoteReplicaGroupPoller() {
+      dataNodeTrackers = new ArrayList<>();
+      fillDataNodeTrackers();
+    }
+
+    public List<DataNodeTracker> getDataNodeTrackers() {
+      return dataNodeTrackers;
+    }
+
+    /**
+     * We will create Data node trackers for every data node
+     * Every tracker will have set active groups and standby groups
+     * Active groups will have {@link #maxReplicaCountPerRequest} maximum number of preassigned replicas
+     * For each data node tracker, we will have only one stand by group
+     * Across all datanode tracker group ids will be unique
+     */
+    private void fillDataNodeTrackers() {
+      Map<DataNodeId, List<RemoteReplicaInfo>> dataNodeToRemoteReplicaInfo = selectReplicas(getRemoteReplicaInfos());
+
+      int currentStartGroupId = 0;
```

Comment:
Will the call for fillDataNodeTrackers() is sequential always ? Otherwise we need to define this variable as atomic.

Reply:
this will be sequential and is thread safe.